pkg/R/RDD.R

```diff
@@ -1323,6 +1323,76 @@ setMethod("zipWithIndex",
            lapplyPartitionsWithIndex(x, partitionFunc)
          })
 
+#' Zip an RDD with another RDD.
+#'
+#' Zips this RDD with another one, returning key-value pairs with the
+#' first element in each RDD second element in each RDD, etc. Assumes
+#' that the two RDDs have the same number of partitions and the same
+#' number of elements in each partition (e.g. one was made through
+#' a map on the other).
+#'
+#' @param x An RDD to be zipped.
+#' @param other Another RDD to be zipped.
+#' @return An RDD zipped from the two RDDs.
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd1 <- parallelize(sc, 0:4)
+#' rdd2 <- parallelize(sc, 1000:1004)
+#' collect(zipRDD(rdd1, rdd2))
+#' # list(list(0, 1000), list(1, 1001), list(2, 1002), list(3, 1003), list(4, 1004))
+#'}
+#' @rdname zipRDD
+#' @aliases zipRDD,RDD
+setMethod("zipRDD",
+          signature(x = "RDD", other = "RDD"),
+          function(x, other) {
+            n1 <- numPartitions(x)
+            n2 <- numPartitions(other)
+            if (n1 != n2) {
+              stop("Can only zip RDDs which have the same number of partitions.")
+            }
+
+            if (x@env$serialized != other@env$serialized || x@env$serialized) {
+              # Even if any RDD is serialized, we need to reserialize it
+              # to make sure its partitions are encoded as a single byte array.
+              # For example, partitions of an RDD generated from partitionBy() may be
+              # encoded as multiple byte arrays.
+              x <- reserialize(x)
+              other <- reserialize(other)
```

Comment:
Do you think we re-serialize the result RDD in partitionBy() so that all serialized RDDs have partitions encoded as a single byte array, so that logic here can be eliminated?